src/bin/pg_autoctl/parsing.c

```diff
@@ -679,3 +681,172 @@ buildPostgresURIfromPieces(URIParams *uriParams, char *pguri)
+
+	return true;
+}
+
+
+/*
+ * parseNodesArrayFromFile parses a Nodes Array from a JSON file, that contains
+ * an array of JSON object with the following properties: node_id, node_lsn,
+ * node_host, node_name, node_port, and potentially node_is_primary.
+ */
+bool
+parseNodesArray(const char *nodesJSON,
+				NodeAddressArray *nodesArray,
+				int nodeId)
+{
+	JSON_Value *json = NULL;
+	JSON_Array *jsArray = NULL;
+	JSON_Value *template =
+		json_parse_string("[{"
+						  "\"node_id\": 0,"
+						  "\"node_lsn\": \"\","
+						  "\"node_name\": \"\","
+						  "\"node_host\": \"\","
+						  "\"node_port\": 0,"
+						  "\"node_is_primary\": false"
+						  "}]");
+	int len = -1;
+	int nodesArrayIndex = 0;
+	int primaryCount = 0;
+
+	json = json_parse_string(nodesJSON);
+
+	/* validate the JSON input as an array of object with required fields */
+	if (json_validate(template, json) == JSONFailure)
+	{
+		log_error("Failed to parse nodes array which is expected "
+				  "to contain a JSON Array of Objects with properties "
+				  "[{node_id:number, node_name:string, "
+				  "node_host:string, node_port:number, node_lsn:string, "
+				  "node_is_primary:boolean}, ...]");
+		return false;
+	}
+
+	jsArray = json_value_get_array(json);
+	len = json_array_get_count(jsArray);
+
+	if (NODE_ARRAY_MAX_COUNT < len)
+	{
+		log_error("Failed to parse nodes array which contains "
+				  "%d nodes: pg_autoctl supports up to %d nodes",
+				  len,
+				  NODE_ARRAY_MAX_COUNT);
+		return false;
+	}
+
+	nodesArray->count = len;
+
+	for (int i = 0; i < len; i++)
+	{
+		NodeAddress *node = &(nodesArray->nodes[nodesArrayIndex]);
+		JSON_Object *jsObj = json_array_get_object(jsArray, i);
+
+		int jsNodeId = (int) json_object_get_number(jsObj, "node_id");
+		uint64_t lsn = 0;
+
+		/* we install the keeper.otherNodes array, so skip ourselves */
+		if (jsNodeId == nodeId)
+		{
+			--(nodesArray->count);
+			continue;
+		}
+
+		node->nodeId = jsNodeId;
+
+		strlcpy(node->name,
+				json_object_get_string(jsObj, "node_name"),
+				sizeof(node->name));
+
+		strlcpy(node->host,
+				json_object_get_string(jsObj, "node_host"),
+				sizeof(node->host));
+
+		node->port = (int) json_object_get_number(jsObj, "node_port");
+
+		strlcpy(node->lsn,
+				json_object_get_string(jsObj, "node_lsn"),
+				sizeof(node->lsn));
+
+		if (!parseLSN(node->lsn, &lsn))
+		{
+			log_error("Failed to parse nodes array LSN value \"%s\"", node->lsn);
+			return false;
+		}
+
+		node->isPrimary = json_object_get_boolean(jsObj, "node_is_primary");
+
+		if (node->isPrimary)
+		{
+			++primaryCount;
+
+			if (primaryCount > 1)
+			{
+				log_error("Failed to parse nodes array: more than one node "
+						  "is listed with \"node_is_primary\" true.");
+				return false;
+			}
+		}
+
+		++nodesArrayIndex;
+	}
+
+	/* now ensure the array is sorted by nodeId */
+	(void) pg_qsort(nodesArray->nodes,
+					nodesArray->count,
+					sizeof(NodeAddress),
+					nodeAddressCmpByNodeId);
```

Comment:
Probably good to validate that nodeIds are unique at this point, given they're sorted here that should be easy.

Reply:
Done now, good point.